MantidPlot/src/ProjectRecoveryGUIs/ProjectRecoveryPresenter.cpp

```diff
@@ -0,0 +1,190 @@
+  m_failureView = nullptr;
+  m_model = new ProjectRecoveryModel(projectRecovery, this);
+  m_openView = RecoveryView;
+  m_startMantidNormallyCalled = false;
+}
+
+ProjectRecoveryPresenter::ProjectRecoveryPresenter(
+    const ProjectRecoveryPresenter &obj) {
+  /// Copy constructor can only copy the Model
+  m_recView = obj.m_recView;
+  m_failureView = obj.m_failureView;
+  m_model = new ProjectRecoveryModel(nullptr, this);
+  *m_model = *obj.m_model;
+  m_mainWindow = obj.m_mainWindow;
+  m_openView = obj.m_openView;
+  m_startMantidNormallyCalled = obj.m_startMantidNormallyCalled;
+}
+
+ProjectRecoveryPresenter::~ProjectRecoveryPresenter() {
+  delete m_recView;
+  delete m_failureView;
+  delete m_model;
+}
+
+bool ProjectRecoveryPresenter::startRecoveryView() {
+  try {
+    m_recView = new ProjectRecoveryView(m_mainWindow, this);
+    m_openView = RecoveryView;
+    m_recView->exec();
+  } catch (...) {
+    return true;
+  }
+
+  // If start mantid normally was called we want to cancel
+  if (m_startMantidNormallyCalled) {
+    return false;
+  }
+
+  // If run has failed and recovery is not running
+  if (m_model->getFailedRun()) {
+    return true;
+  }
+  return false;
+}
+
+bool ProjectRecoveryPresenter::startRecoveryFailure() {
+  try {
+    m_failureView = new RecoveryFailureView(m_mainWindow, this);
+    m_openView = FailureView;
+    m_failureView->exec();
+  } catch (...) {
+    return true;
+  }
+
+  // If start mantid normally was called we want to cancel
+  if (m_startMantidNormallyCalled) {
+    return false;
+  }
+
+  // If run has failed and recovery is not running
+  if (m_model->getFailedRun()) {
+    return true;
+  }
+  return false;
+}
+
+QStringList ProjectRecoveryPresenter::getRow(int i) {
+  auto vec = m_model->getRow(i);
+  QStringList returnVal;
+  for (auto i = 0; i < 3; ++i) {
+    QString newString = QString::fromStdString(vec[i]);
+    returnVal << newString;
+  }
+  return returnVal;
+}
+
+void ProjectRecoveryPresenter::recoverLast() {
+  if (m_model->hasRecoveryStarted())
+    return;
+  auto checkpointToRecover = m_model->decideLastCheckpoint();
+  m_model->recoverSelectedCheckpoint(checkpointToRecover);
+}
+
+void ProjectRecoveryPresenter::openLastInEditor() {
+  if (m_model->hasRecoveryStarted())
+    return;
+  auto checkpointToRecover = m_model->decideLastCheckpoint();
+  m_model->openSelectedInEditor(checkpointToRecover);
+}
+
+void ProjectRecoveryPresenter::startMantidNormally() {
+  m_startMantidNormallyCalled = true;
+  m_model->startMantidNormally();
+}
+
+void ProjectRecoveryPresenter::recoverSelectedCheckpoint(QString &selected) {
+  if (m_model->hasRecoveryStarted())
+    return;
+  auto checkpointToRecover = selected.toStdString();
+  m_model->recoverSelectedCheckpoint(checkpointToRecover);
+}
+
+void ProjectRecoveryPresenter::openSelectedInEditor(QString &selected) {
+  if (m_model->hasRecoveryStarted())
+    return;
+  auto checkpointToRecover = selected.toStdString();
```

Comment:
I'm writing the same suggestion comment here a third time, I think. Why the code duplication?

Reply:
Because they different member functions represent different calls from different parts of the Two views that are represented by this Presenter, each one happens at a different button press etc. So for example this gets a QString which needs to be a stdString which is already only one call, so reducing it isn't really possible without being pretty extreme.

Reply:
Sounds reasonable.